### PR TITLE
[WIP] Fix handling of ECMA-335 file format corner case

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
This fixes the handling of a corner case in the ECMA-335 spec in which a module that has 0xFFFF methods *and* the last type in the module has no methods. (I'm going to refer to methods, but this also impacts fields, and to a lesser extent properties and events.)

The file format represents "last type with no methods" as a type whose methods start at the index of Number_Of_Methods in the assembly. For a module with exactly 0xFFFF methods, this index would be 0x10000 (index zero is reserved and not assigned to an actual method). The file format also says that the width of the database column that specifies method indices is 2 bytes if the number of methods is less or equal to 0xFFFF. You probably see where things go wrong.

The spec also says that the "type with no methods" situation can be expressed as "methods of this type start at index 0" (the invalid index), but CLR didn't implement it this way, so it's unlikely anyone emits them like that.

This was famously hit in the Windows build lab where we were forced to ask people to add a dummy method as a workaround (https://twitter.com/h0x0d/status/1012176973433409536). After giving the workaround, I thought - "what a great corner case!", filed a bug to fix this, checked with the rest of the team where nobody remembered anyone hitting this since the v1 of the CLR, and moved on with life. A couple months later someone else hit this. Since this is too big of a coincidence, here's my attempt to fix the corner case.

This pull request has 3 parts, with increasing levels of complexity:

* Don't stomp over HRESULT if metadata saving operation failed. This was causing ILASM to generate broken assemblies when this situation is hit, without an error. This is the least problematic part of this pull request and on it's own, will at least make ILASM fail.
* Support in the metadata writer to emit index zero if we're in the corner case situation (this hopefully avoids tooling breakage - we could emit index zero for other last-type-has-no-methods situations, but since nobody can handle that, we limit it to the case where the output would be broken anyways)
* Support in the metadata reader for types whose method list starts at index zero.

If we go with fixing this, we'll also need to have a look at other tooling, e.g. dotnet/corefx#29554 that impacts Roslyn and a bunch of other tools.

### CODE REVIEW

File: src/md/compiler/import.cpp [hunk 1]
```diff
@@ -1342,19 +1327,17 @@ STDMETHODIMP RegMeta::EnumProperties(         // S_OK, S_FALSE, or error.
...
         {
-            IfFailGo(m_pStgdb->m_MiniMd.GetPropertyMapRecord(ridPropertyMap, &pPropertyMapRec));
-            ridStart = pMiniMd->getPropertyListOfPropertyMap(pPropertyMapRec);
+            IfFailGo(pMiniMd->getStartPropertyListOfPropertyMap(ridPropertyMap, &ridStart));
             IfFailGo(pMiniMd->getEndPropertyListOfPropertyMap(ridPropertyMap, &ridEnd));
-            ridMax = pMiniMd->getCountPropertys() + 1;
-            if(ridStart == 0) ridStart = 1;
-            if(ridEnd > ridMax) ridEnd = ridMax;
-            if(ridStart > ridEnd) ridStart=ridEnd;
+            //ridMax = pMiniMd->getCountPropertys() + 1;
```
Comment (MichalStrehovsky) [inline src/md/compiler/import.cpp]: This is a bunch of suspicious code I'll need to investigate more. Hence the WIP.

File: src/md/ceefilegen/cceegen.cpp [hunk 2]
```diff
@@ -610,7 +610,7 @@ HRESULT CCeeGen::emitMetaData(IMetaDataEmit *emitter, CeeSection* section, DWORD
     
     if (! m_fTokenMapSupported) {
         // Remove the handler that we set
-        hr = emitter->SetHandler(NULL);
+        emitter->SetHandler(NULL);
```
Comment (MichalStrehovsky) [inline src/md/ceefilegen/cceegen.cpp]: This cannot actually fail. Other place in the file that calls this with NULL ignores the HRESULT too.

File: src/md/compiler/import.cpp (hunk 1 above)
Comment (MichalStrehovsky) [reply src/md/compiler/import.cpp]: Welp, as expected, this code predates TFS source history of the CLR. It kind of looks like it was trying to fix some bad tooling. This code would no longer work after the change that prevents `ridStart` from ever be reported as zero, so I propose we just delete this (and other places like this).

**Comment by MichalStrehovsky** [comment]
@davidwr What do you think about this?

Cc @sergiy-k

**Comment by jkotas** [comment]
Nit: It is ECMA-335. ECMA-355 is [this](https://www.ecma-international.org/publications/standards/Ecma-355.htm).

**Comment by steveharter** [comment]
I don't think the last type has to have zero methods, just 0xFFFF total. I'll share a repo offline.

File: src/md/runtime/metamodelro.cpp [hunk 3]
```diff
@@ -179,6 +179,12 @@ CMiniMd::Impl_GetEndRidForColumn(   // The End rid.
         BYTE *pRow;
         IfFailRet(Impl_GetRow(nTableIndex, nRowIndex + 1, &pRow));
         *pEndRid = getIX(pRow, def);
+
+        if (InvalidRid(*pEndRid))
+        {
+            _ASSERTE(nTargetTableIndex < TBL_COUNT);
+            *pEndRid = m_Schema.m_cRecs[nTargetTableIndex] + 1;
```
[the remark this answers is not in the document]
Comment (MichalStrehovsky) [reply src/md/runtime/metamodelro.cpp]: That will be pretty untestable though, unless I update ILASM to emit that. But then nobody but CoreCLR will be able to read CoreCLR's ILASM output.

[the remark this answers is not in the document]
Comment (MichalStrehovsky) [reply src/md/runtime/metamodelro.cpp]: (E.g. the ILDASM with the nice GUI we ship with the desktop CLR dislikes null RID a lot in that position.)

**Comment by davidwrighton** [comment]
In general, I believe this is a reasonable path forward, but I'd like to hear @tmat's opinion. I don't really like a path which leads us to having metadata that the desktop CLR can't read, but I don't see a better way to fix it for things like Reflection.Emit from CoreCLR. OTOH, perhaps we should make this sort of change to the format, and then push up the stack so that Roslyn and other metadata tools that continue to support desktop CLR use an extra hidden method to workaround the issue.

**Comment by MichalStrehovsky** [comment]
> In general, I believe this is a reasonable path forward, but I'd like to hear @tmat's opinion. I don't really like a path which leads us to having metadata that the desktop CLR can't read, but I don't see a better way to fix it for things like Reflection.Emit from CoreCLR

Right, this is why I propose we only ever emit the 0 list for this one specific case where the existing tools would emit a broken assembly - that way we are not worse off than if we didn't do anything for readers that are not enlightened (it's a weird error either way), but readers that are enlightened can do the right thing. For writers, the distinction would be successful writing vs. a writing failure (provided they don't have bugs around this failure mode like ILASM had)

My line of thinking would be to update the low level tooling (S.R.Metadata, CCI, Cecil, etc.) to do the same thing as what I did here (emit zero rid only for the specific corner case, and add support for reading it).

I would think that Roslyn would probably want to just emit an extra method to be super-compatible with old runtimes and tools. This choice would be available to other consumers of the low level tools (they can avoid hitting the corner case emission path by adding an extra thing to the module) - if they do nothing, they get an assembly that only works for enlightened readers (which is an improvement over getting an assembly that doesn't work at all).

But yes, I want to wait for @tmat's opinion as well.

**Comment by tmat** [comment]
I'm in favor of this approach. Would you mind filing an issue on Roslyn if there isn't one open yet, to emit extra method/field/property to work around the corner cases?

**Comment by maryamariyan** [comment]
Thank you for your contribution. As announced in dotnet/coreclr#27549 this repository will be moving to dotnet/runtime on November 13. If you would like to continue working on this PR after this date, the easiest way to move the change to dotnet/runtime is:
1.	In your coreclr repository clone, create patch by running `git format-patch origin`
2.	In your runtime repository clone, apply the patch by running `git apply --directory src/coreclr <path to the patch created in step 1>`

**Comment by maryamariyan** [comment]
Thank you for your contribution. As announced in #27549 the dotnet/runtime repository will be used going forward for changes to this code base. Closing this PR as no more changes will be accepted into master for this repository. If you’d like to continue working on this change please move it to dotnet/runtime.

**Pull request closed without merging**